tests/common.py

```diff
@@ -1352,6 +1354,46 @@ def assert_lists_same(a: list[Any], b: list[Any]) -> None:
         assert i in a
 
 
+async def assert_registries_and_states_for_config_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    snapshot: SnapshotAssertion,
+    *,
+    platform: Platform | None = None,
+) -> None:
+    """Ensure that registry content and state machine match snapshot."""
+    # Ensure devices are correctly registered
+    device_registry = dr.async_get(hass)
+    device_entries = dr.async_entries_for_config_entry(
+        device_registry, config_entry.entry_id
+    )
+    assert device_entries == snapshot
+
+    # Ensure entities are correctly registered, optionally filtered by platform
+    entity_registry = er.async_get(hass)
+    entity_entries = er.async_entries_for_config_entry(
+        entity_registry, config_entry.entry_id
+    )
+    if platform:
+        entity_entries = [ent for ent in entity_entries if ent.domain == platform]
+    assert entity_entries == snapshot
+
+    # If some entities are disabled, enable them and reload before checking states
+    needs_reload = False
+    for ent in entity_entries:
+        if ent.disabled:
+            entity_registry.async_update_entity(ent.entity_id, **{"disabled_by": None})
+            needs_reload = True
+
+    if needs_reload:
+        await hass.config_entries.async_reload(config_entry.entry_id)
+        await hass.async_block_till_done()
+
+    # Ensure entities states are correct
+    for ent in entity_entries:
+        assert hass.states.get(ent.entity_id) == snapshot(name=ent.entity_id)
```

Comment:
It is better to provide the expected entities. The test will pass if an entity isn't in the registry by accident.

(Functional vs snapshot testing).

Reply:
I didn't use `async_all` because that doesn't allow filtering by `config_entry`.
I have adjusted this now to build a list of states, and then compare that with the snapshot.
Is that what you meant?

Reply:
No, that is not what I meant.

If one creates a new entity, that is omitted from the registry, the test will pass.
The expected entities are not checked at this point (which is the goal here primarily?)

Additionally, as per Discord discussion, I think this should not be a helper at all.

Reply:
Yep, I've removed the helper method now.

You are correct that "If one creates a new entity, that is omitted from the registry, the test will pass." but that is already the case I think all tests in all integrations. I don't really see how we can add a test in advance for that scenario.

What these tests do is make sure that entities are not accidently removed or amended.